pallets/parachain-system/src/validate_block/trie_cache.rs

```diff
@@ -0,0 +1,110 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate alloc;
+
+use core::cell::RefCell;
+use hash_db::{HashDB, Hasher};
+use hashbrown::hash_map::{Entry, HashMap};
+use sp_state_machine::{TrieBackendStorage, TrieCacheProvider};
+use sp_std::boxed::Box;
+use sp_trie::NodeCodec;
+use trie_db::{node::NodeOwned, TrieCache, TrieError};
+
+/// Special purpose trie cache implementation that is able to cache an unlimited number
+/// of values. To be used in `validate_block` to serve values and nodes that
+/// have already been loaded and decoded from the storage proof.
+pub(crate) struct SimpleTrieCache<'a, H: Hasher> {
+	node_cache: core::cell::RefMut<'a, HashMap<H::Out, NodeOwned<H::Out>>>,
+	value_cache: core::cell::RefMut<'a, HashMap<Box<[u8]>, trie_db::CachedValue<H::Out>>>,
+}
+
+impl<'a, H: Hasher> trie_db::TrieCache<NodeCodec<H>> for SimpleTrieCache<'a, H> {
+	fn lookup_value_for_key(&mut self, key: &[u8]) -> Option<&trie_db::CachedValue<H::Out>> {
+		self.value_cache.get(key)
+	}
+
+	fn cache_value_for_key(&mut self, key: &[u8], value: trie_db::CachedValue<H::Out>) {
+		self.value_cache.insert(key.into(), value);
+	}
+
+	fn get_or_insert_node(
+		&mut self,
+		hash: <NodeCodec<H> as trie_db::NodeCodec>::HashOut,
+		fetch_node: &mut dyn FnMut() -> trie_db::Result<
+			NodeOwned<H::Out>,
+			H::Out,
+			<NodeCodec<H> as trie_db::NodeCodec>::Error,
+		>,
+	) -> trie_db::Result<&NodeOwned<H::Out>, H::Out, <NodeCodec<H> as trie_db::NodeCodec>::Error> {
+		match self.node_cache.entry(hash) {
+			Entry::Occupied(entry) => Ok(entry.into_mut()),
+			Entry::Vacant(entry) => {
+				let fetched = match fetch_node() {
+					Ok(new_node) => new_node,
+					Err(e) => return Err(e),
+				};
+				Ok(entry.insert(fetched))
+			},
+		}
+	}
+
+	fn get_node(
+		&mut self,
+		hash: &H::Out,
+	) -> Option<&NodeOwned<<NodeCodec<H> as trie_db::NodeCodec>::HashOut>> {
+		self.node_cache.get(hash)
+	}
+}
+
+/// Provider of [`SimpleTrieCache`] instances.
+pub(crate) struct CacheProvider<H: Hasher> {
+	node_cache: RefCell<HashMap<H::Out, NodeOwned<H::Out>>>,
+	value_cache: RefCell<HashMap<Box<[u8]>, trie_db::CachedValue<H::Out>>>,
+}
+
+impl<H: Hasher> CacheProvider<H> {
+	/// Constructs a new instance of CacheProvider with an uninitialized state
+	/// and empty node and value caches.
+	pub fn new() -> Self {
+		CacheProvider { node_cache: Default::default(), value_cache: Default::default() }
+	}
+}
+
+impl<H: Hasher> TrieCacheProvider<H> for CacheProvider<H> {
+	type Cache<'a> = SimpleTrieCache<'a, H> where H: 'a;
+
+	fn as_trie_db_cache(&self, _storage_root: <H as Hasher>::Out) -> Self::Cache<'_> {
+		SimpleTrieCache {
+			value_cache: self.value_cache.borrow_mut(),
+			node_cache: self.node_cache.borrow_mut(),
+		}
+	}
+
+	fn as_trie_db_mut_cache(&self) -> Self::Cache<'_> {
+		SimpleTrieCache {
+			value_cache: self.value_cache.borrow_mut(),
+			node_cache: self.node_cache.borrow_mut(),
+		}
+	}
```

Comment:
This function is called when we call `storage_root`. The `values` that are being created while calculating the `storage_root` should not be added to the `value_cache` as otherwise they may would overwrite some cached value from the "real trie" (we only distinguish using the key and the key stays the same in each trie). If another operation then would want to fetch this overwritten value, it would get some invalid data.

The nodes are accessed by their hash, so that should be no problem. However, the question in general is if we even need to cache these nodes. These nodes will never be accessible from the "storage root" we are operating on.

Reply:
The value cache is now disabled when building storage roots. Due to how `get_or_insert_node` works I think it does still make sense to take ownership of the fetched node and cache it.